packages/api/src/cli/codegen/languages/typescript.ts

```diff
@@ -383,85 +376,103 @@ sdk.server('https://eu.api.example.com/v14');`)
+      declarations: [
+        {
+          name: 'createSDK',
+          initializer: writer => {
+            // `ts-morph` doesn't have any way to cleanly create an IFEE.
+            writer.writeLine('(() => { return new SDK(); })()');
+            return writer;
+          },
+        },
+      ],
     });
 
-    if (this.outputJS) {
-      return this.project
-        .emitToMemory()
-        .getFiles()
-        .map(sourceFile => ({
-          [path.basename(sourceFile.filePath)]: TSGenerator.formatter(sourceFile.text),
-        }))
-        .reduce((prev, next) => Object.assign(prev, next));
-    }
+    sourceFile.addExportAssignment({ isExportEquals: false, expression: 'createSDK' });
 
-    return [
-      ...this.project.getSourceFiles().map(sourceFile => ({
-        [sourceFile.getBaseName()]: TSGenerator.formatter(sourceFile.getFullText()),
-      })),
+    return sourceFile;
+  }
 
-      // Because we're returning the raw source files for TS generation we also need to separately
-      // emit out our declaration files so we can put those into a separate file in the installed
-      // SDK directory.
-      ...this.project
-        .emitToMemory({ emitOnlyDtsFiles: true })
-        .getFiles()
-        .map(sourceFile => ({
-          [path.basename(sourceFile.filePath)]: TSGenerator.formatter(sourceFile.text),
-        })),
-    ].reduce((prev, next) => Object.assign(prev, next));
+  /**
+   * Create our main schemas file. This is where all of the JSON Schema that our TypeScript typing
+   * infrastructure sources its data from. Without this there are no types.
+   *
+   */
+  createSchemasFile() {
+    const sourceFile = this.project.createSourceFile('schemas.ts', '');
+
+    const sortedSchemas = new Map(Array.from(Object.entries(this.schemas)).sort());
+
+    Array.from(sortedSchemas).forEach(([schemaName, schema]) => {
+      sourceFile.addVariableStatement({
+        declarationKind: VariableDeclarationKind.Const,
+        declarations: [
+          {
+            name: schemaName,
+            initializer: writer => {
+              /**
+               * This is the conversion prefix that we add to all `$ref` pointers we find in
+               * generated JSON Schema.
+               *
+               * Because the pointer name is a string we want to have it reference the schema
+               * constant we're adding into the codegen'd schema file. As there's no way, not even
+               * using `eval()` in this case, to convert a string to a constant we're prefixing
+               * them with this so we can later remove it and rewrite the value to a literal.
+               * eg. `'Pet'` becomes `Pet`.
+               *
+               * And because our TypeScript type name generator properly ignores `:`, this is safe
+               * to prepend to all generated type names.
+               */
+              let str = JSON.stringify(schema);
+              str = str.replace(/"::convert::([a-zA-Z_$\\d]*)"/g, '$1');
```

Comment:
I don't like this but it works. We do similar things in HTTPSnippet for turning some strings into literals for snippets. https://github.com/readmeio/httpsnippet/blob/main/src/targets/node/fetch/client.ts#L149-L151